Use allow-list for APM agent settings

mosche opened the pull request:
Prevent invalid and miss-configuration of APM agent using an explicit allow-list of setting keys.
Additionally this removes the redundant `APM_AGENT_DEFAULT_SETTINGS` in favor of a single definition of defaults in `APMJvmOptions`.
(Closes ES-6916)

mosche committed:
Use allow-list for APM agent settings (ES-6916)

Prevent invalid and miss-configuration of APM agent using an explicit allow-list of setting keys.

(Closes ES-6916)

mosche commented on modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:
```diff
@@ -114,15 +115,130 @@ public void setAgentSetting(String key, String value) {
+        "classes_excluded_from_instrumentation",
+        "trace_methods",
+        "trace_methods_duration_threshold",
+        "central_config", // TODO TBD disabled in APMJvmOptions defaults, does this impact usage of config_file?
```
Not sure about this one, could it impact usage of / conflict with `config_file`?

stu-elastic replied:
Let's remove this one as it impacts `config_file`.

mosche replied:
👍

mosche committed:
Consistent configuration of APM agent default settings in APMJvmOptions.

Move configuration defaults from APMAgentSettings to APMJvmOptions to consistently apply these in one location only.

mosche commented on modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:
```diff
@@ -114,15 +115,130 @@ public void setAgentSetting(String key, String value) {
+        "sanitize_field_names",
+        "enable_instrumentations",
+        "disable_instrumentations",
+        "enable_experimental_instrumentations", // TODO TBD allow disabling this, conflicts with APMJvmOptions
```
Here's another PR to apply defaults in a single consistent location:
https://github.com/elastic/elasticsearch/pull/104149/

stu-elastic replied:
Yeah it looks like `APM_AGENT_DEFAULT_SETTINGS` should be nuked.

mosche replied:
👍 I dropped `APM_AGENT_DEFAULT_SETTINGS` and removed `enable_experimental_instrumentations` from the allow list.

mosche committed:
fix checkstyle

stu-elastic commented on modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:
```diff
@@ -114,15 +115,130 @@ public void setAgentSetting(String key, String value) {
+        "data_flush_timeout",
+
+        // Stacktraces:
+        "application_packages", // TODO TBD default configured by APMJvmOptions, allow overwrites?
```
We can allow it for now.

mosche replied:
👍

stu-elastic commented on modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:
```diff
@@ -132,7 +248,7 @@ public void setAgentSetting(String key, String value) {
             final String defaultValue = APM_AGENT_DEFAULT_SETTINGS.getOrDefault(key, "");
             return new Setting<>(qualifiedKey, defaultValue, (value) -> {
-                if (PROHIBITED_AGENT_KEYS.contains(key)) {
+                if (qualifiedKey.equals("_na_") == false && PERMITTED_AGENT_KEYS.contains(key) == false) {
```
I think the `_na_` had to do with `Setting.affixKeySetting`'s `delegateFactory` which is used via `Setting.prefixKeySetting` by `APM_AGENT_SETTINGS`.  Are you sure it's ok to remove this check?

mosche replied:
Moving from deny to allow list, I had to explicitly allow `_na_`. As you mentioned the `delegateFactory` is always invoked once with `_na_`, which obviously shouldn't fail.

mosche committed:
Consistent configuration of APM agent default settings in APMJvmOptions.

Move configuration defaults from APMAgentSettings to APMJvmOptions to consistently apply these in one location only.

mosche committed:
feedback

mosche committed:
Test that forbidden / unknown settings are rejected

pgomulka commented on build-tools/src/main/java/org/elasticsearch/gradle/testclusters/RunTask.java:
```diff
@@ -213,8 +213,8 @@ public void beforeStart() {
-                } else if (node.getSettingKeys().contains("tracing.apm.agent.enabled") == false) { // tracing
-                    node.setting("tracing.apm.agent.enable", "false");
+                } else if (node.getSettingKeys().contains("tracing.apm.enabled") == false) { // tracing
+                    node.setting("tracing.apm.enable", "false");
```
turns out this is not an affix seting and we could deprecate it and use 'telemetry.tracing.enabled' right?

mosche replied:
Yes, I noticed that when talking to Stu, I'll work on that next 👍

pgomulka reviewed (commented):
I left few comments. I think they are not really related to the exact thing in this PR (allow-list of settings), but perhaps we could follow up in a separate PR

mosche committed:
Fix test names

mosche commented on modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettings.java:
```diff
@@ -114,26 +94,140 @@ public void setAgentSetting(String key, String value) {
     private static final String APM_SETTING_PREFIX = "tracing.apm.";
```
Can we postpone that for a later PR when cleaning up metric names? Likely that will touch all of this anyways.

mosche commented on modules/apm/src/test/java/org/elasticsearch/telemetry/apm/internal/APMAgentSettingsTests.java:
```diff
@@ -40,48 +40,29 @@ public void test_whenTracerDisabled_setsRecordingProperty() {
      */
-    public void test_whenTracerCreated_clusterSettingsAlsoApplied() {
-        APMAgentSettings apmAgentSettings = spy(new APMAgentSettings());
+    public void test_invalidSettingsAreRejected() {
```
Done

mosche commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into ES-6916_apm_agent_settings_whitelist

pgomulka reviewed (approved):
LGTM, thank you
